### PR TITLE
Add a way for business logic to throw an expcetion that is propagated to the API.

### PR DESCRIPTION
**Pull request by tomprince**
This is a followup to #405, which allows code to throw errors that are passed to users.

### CODE REVIEW

File: newsfragments/430.minor [hunk 1]
```diff
@@ -0,0 +1 @@
+Add your info here
```
Comment (meejah) [inline newsfragments/430.minor:1]: We should be following the dev-guide and filing tickets, _then_ doing PRs for those tickets. If you want to propose changes, please propose a change to the dev guide.

https://github.com/LeastAuthority/magic-folder/blob/main/docs/development.rst

**Comment by meejah** [comment]
I see a bunch of Nix changes; are those a mistake?

File: newsfragments/430.minor (hunk 1 above)
Comment (tomprince) [reply newsfragments/430.minor:1]: This is part of #210, but split out into a separate PR to be easier to review.

**Comment by tomprince** [comment]
> I see a bunch of Nix changes; are those mistakenly included in this PR?

`@attr.s(auto_exc=...)` needs a newer attrs than is in the nix snapshot we are using.

File: newsfragments/430.minor (hunk 1 above)
Comment (meejah) [reply newsfragments/430.minor:1]: that's fine, but having PRs point newsfragments at themselves isn't really the point. i get that it's annoying sometimes, but that's what we've got ...

File: src/magic_folder/common.py [hunk 2]
```diff
@@ -33,6 +35,29 @@ class BadMetadataResponse(Exception):
...
+    """
+
+    reason = attr.ib(validator=attr.validators.instance_of(unicode))
+    code = attr.ib(
+        default=None,
+        validator=attr.validators.optional(attr.validators.instance_of(int)),
+    )
+
+    @classmethod
+    def from_exception(cls, code, exception, prefix=None):
```
Comment (meejah) [inline src/magic_folder/common.py:51]: should document the params. I assume "code" is an HTTP status code, but would be good to say that (if it's true).

File: src/magic_folder/test/test_web.py [hunk 3]
```diff
@@ -986,6 +987,66 @@ def test_add_participant(self, folder_name, personal_dmd):
...
+        local_path = FilePath(self.mktemp())
+        local_path.makedirs()
+        folder_config = magic_folder_config(
+            create_local_author(author),
+            local_path,
+        )
+
+        root = create_fake_tahoe_root()
+        # put our Collective DMD into the fake root
+        root._uri.data[folder_config["collective-dircap"]] = dumps([
```
Comment (meejah) [inline src/magic_folder/test/test_web.py:1007]: For future reference, there's `root.add_data` and `root.add_mutable_data` that don't use internals

File: src/magic_folder/web.py [hunk 4]
```diff
@@ -264,41 +280,38 @@ def add_participant(self, request, folder_name):
...
-            if not IDirnodeURI.providedBy(dmd):
-                raise _InputError("personal_dmd must be a directory-capability")
-            if not dmd.is_readonly():
-                raise _InputError("personal_dmd must be read-only")
-            personal_dmd_cap = participant["personal_dmd"]
-        except _InputError as e:
-            request.setResponseCode(http.BAD_REQUEST)
-            returnValue(json.dumps({"reason": str(e)}))
+        dmd = tahoe_uri_from_string(participant["personal_dmd"])
+        if not IDirnodeURI.providedBy(dmd):
```
Comment (meejah) [inline src/magic_folder/web.py:310]: I know this is just re-indented existing code, but there's `util.capabilitiies.*` methods (for future reference).

**Review by meejah — COMMENTED** [review]
Code looks fine, but @exarkun will have to review the Nix changes (I guess maybe more Nix changes are required still though?)

File: src/magic_folder/test/test_web.py (hunk 3 above)
Comment (tomprince) [reply src/magic_folder/test/test_web.py:1007]: Hmm. I went and looked a the other tests, and it looks like they aren't using those either. I think there may be a mismatch between the API the fake provides, and what is useful for us. I suspect we should add some wrappers or extend it to have a more friendly API. I filed #436 to do that.

**Comment by tomprince** [comment]
> will have to review the Nix changes (I guess maybe more Nix changes are required still though?

I think the nix changes are fine without review (once #432 lands so they work), since it is just bumping the versions being used to something closer to what is used on other platforms.

**Pull request merged**